File: leetcode/947/947. Most Stones Removed with Same Row or Column.go
package leetcode

type QuickUnion struct {
	id []int
	sz []int //counts number of elements in the tree rooted at i.
}

func makeQuickUnion(size int) *QuickUnion {
	id := make([]int, size, size)
	sz := make([]int, size, size)

	for index := 0; index < size; index++ {
		id[index] = index
		sz[index] = 1
	}

	return &QuickUnion{id: id, sz: sz}
}

func (q *QuickUnion) root(i int) int {
	for i != q.id[i] {
		q.id[i] = q.id[q.id[i]] //for path compression
		i = q.id[i]
	}

	return i
}

func (q *QuickUnion) find(node1 int, node2 int) bool {
	return q.root(node1) == q.root(node2)
}

func (q *QuickUnion) unite(node1 int, node2 int) {
	root1 := q.root(node1)
	root2 := q.root(node2)

	if q.sz[root1] < q.sz[root2] {
		q.id[root1] = root2
		q.sz[root2] += q.sz[root1]
	} else {
		q.id[root2] = root1
		q.sz[root1] += q.sz[root2]
	}
}

func (q *QuickUnion) treeCount() int {
	treeCount := 0
	for i := 0; i < len(q.id); i++ {
		if q.id[i] == i {
			treeCount++
		}
	}

	return treeCount
}

func removeStones(stones [][]int) int {
	size := len(stones)
	q := makeQuickUnion(size)

	for i := 0; i < size; i++ {
		//no point to iterate from 0 again, because the
		//previous i points were already covered
		for j := i + 1; j < size; j++ {
			if stones[i][0] == stones[j][0] || stones[i][1] == stones[j][1] {
				q.unite(i, j)
			}
		}
	}
	//count #trees aka connected components
	return size - q.treeCount()
}

File: leetcode/947/947. Most Stones Removed with Same Row or Column.py
class DSU:
    def __init__(self, N):
        self.p = [n for n in range(N)]

    def find(self, x):
        if self.p[x] != x:
            self.p[x] = self.find(self.p[x])
        return self.p[x]

    def union(self, x, y):
        xr = self.find(x)
        yr = self.find(y)
        self.p[xr] = yr

class Solution(object):
    def removeStones(self, stones):
        N = len(stones)
        dsu = DSU(20000)

        for x, y in stones:
            dsu.union(x, y + 10000)
            print('union(',x,',',y+10000,')')
        
        l = set(dsu.find(x) for x, _ in stones)
        print(l)
        return N - len(l)

s = Solution()
print(s.removeStones([[0, 0], [0, 2], [1, 1], [2, 0], [2, 2]]))